splitio/tasks/util/workerpool.py

```diff
@@ -134,3 +134,99 @@ def _wait_workers_shutdown(self, event):
             for worker_event in self._worker_events:
                 worker_event.wait()
             event.set()
+
+
+class WorkerPoolAsync(object):
+    """Worker pool async class to implement single producer/multiple consumer."""
+
+    _abort = object()
+
+    def __init__(self, worker_count, worker_func):
+        """
+        Class constructor.
+
+        :param worker_count: Number of workers for the pool.
+        :type worker_func: Function to be executed by the workers whenever a messages is fetched.
+        """
+        self._semaphore = asyncio.Semaphore(worker_count)
+        self._queue = asyncio.Queue()
+        self._handler = worker_func
+        self._aborted = False
+        self._failed = False
+
+    async def _schedule_work(self):
+        """wrap the message handler execution."""
+        while True:
+            message = await self._queue.get()
+            if message == self._abort:
+                self._aborted = True
+                return
+            asyncio.get_running_loop().create_task(self._do_work(message))
+
+    async def _do_work(self, message):
+        """process a single message."""
+        try:
+            await self._semaphore.acquire() # wait until "there's a free worker"
+            if self._aborted: # check in case the pool was shutdown while we were waiting for a worker
+                return
+            await self._handler(message._message)
+        except Exception:
+            _LOGGER.error("Something went wrong when processing message %s", message)
+            _LOGGER.debug('Original traceback: ', exc_info=True)
+            self._failed = True
```

Comment:
i don't think `self._failed` will work, since it's signaling that something failed when processing one message, which may or may not be from the same batch, and since this is async, it can become somewhat unpredictable.

how do you feel about adding a `failed` property to the `TaskCompletionWrapper` instead (and maybe a method to `BatchCompletionHandler` as well, that checks the status of nested items to look for possible errors? that way we can track what failed (it can be a bool, the expcetion's message or both, follow your heart).

ideally we can consume the WP's interface like this:
```
       jobs = await self._worker_pool.submit_work(segment_names)
        if dont_wait:
            return True
        await jobs.await_completion()
        return jobs.all_successful()
```

Reply:
fixed